src/window.cpp

```diff
@@ -876,60 +876,98 @@ static void DispatchMouseWheelEvent(Window *w, NWidgetCore *nwid, int wheel)
  * @param bottom Bottom edge of the rectangle that should be repainted
  * @param gfx_dirty Whether to mark gfx dirty
  */
-void DrawOverlappedWindow(Window *w, int left, int top, int right, int bottom, bool gfx_dirty)
-{
-	const Window *v;
-	FOR_ALL_WINDOWS_FROM_BACK_FROM(v, w->z_front) {
-		if (MayBeShown(v) &&
-				right > v->left &&
-				bottom > v->top &&
-				left < v->left + v->width &&
-				top < v->top + v->height) {
-			/* v and rectangle intersect with each other */
-			int x;
-
-			if (left < (x = v->left)) {
-				DrawOverlappedWindow(w, left, top, x, bottom, gfx_dirty);
-				DrawOverlappedWindow(w, x, top, right, bottom, gfx_dirty);
-				return;
+void DrawOverlappedWindow(Window *w, const int left, const int top, const int right, const int bottom, const bool gfx_dirty)
+{
+	std::function<void(Window const * const, int, int, int, int)> draw_overlapped;
+	draw_overlapped = [&w, &gfx_dirty, &draw_overlapped](Window const * const w_from, int left, int top, int right, int bottom){
```

Comment:
In general I would be wary of this sort of structure.
std::function is really for dynamic dispatch, as it adds significant overhead in code size, run time and memory use.
The call targets are entirely static here, just calling the target function directly as before is fine.

Reply:
How else would you do a nested recursive function in C++?
The entire point of the closure is to be able to eliminate unnecessary global variables (well, not so much in this case; but in general).

Reply:
Wrap the relevant code in a class. The class can even be private to the compilation unit and have a simple function as call interface in the header.

Reply:
I'd say that if this much code/overhead is necessary to eliminate a global variable, then the global variable is fine. Hell, I think a goto would be better than this.

Reply:
Why, oh why, doesn't C just support nested functions like any other decent language? It's only a layer of indirection via (a copy of) the parent's base pointer, after all.

True, C++ has closures and lambdas but, as seen above, the current implementation (C++11) still misses the mark for this one incredibly useful feature included in nearly every other language :disappointed:.